### PR TITLE
Adds PULP timer_unit driver for timer0

### PR DESCRIPTION
**Pull request by andstepan**
This is the driver implements the interfacing for the timer peripheral found in the headsail-vp. The peripheral found in the VP is not the PULP APB Timer, as specified in the Headsail documentation, but is the PULP Timer Unit ([https://github.com/pulp-platform/timer_unit](https://github.com/pulp-platform/timer_unit)). This means, that **this driver is valid for the VP ONLY**.

Summary of changes:
examples/headsail-bp/Cargo.toml: Added timer_unit module dependencies and example entry
examples/headsail-bp/src/lib.rs: Added timer_unit module
examples/headsail-bp/src/mmap.rs: Added timer0 address constant
examples/headsail-bp/src/timer_unit.rs: Driver implementation (New file)
examples/hpc/hello-hpc/examples/timer0.rs: Example implementation (New file)

### CODE REVIEW

**Comment by vilukissa68** [comment]
For some reason @andstepan doesn't have rights to run the CI, so I approved this run.

**Comment by vilukissa68** [comment]
Tested example with:
```sh
cargo run --example timer0 --target riscv64imac-unknown-none-elf
```
Works well!

File: examples/headsail-bsp/Cargo.toml
Comment (hegza) [inline examples/headsail-bsp/Cargo.toml:58]: This entry implies that there should be an example under project `headsail-bsp` at location `examples/headsail-bsp/examples/timer0.rs". In reality, the example is located under project `hello-hpc` at `examples/hpc/hello-hpc/examples/timer0.rs`.

Pick one:

- Move the exercise from hello-hpc/examples to headsail-bsp/examples (**preferred**)
- Move the [[example]] entry from headsail-bsp/Cargo.toml to hello-hpc/Cargo.toml

File: examples/headsail-bsp/src/lib.rs
Comment (hegza) [inline examples/headsail-bsp/src/lib.rs:6]: Since there may be several timers on the platform, an argument could be made that being very specific with the name might be a good idea, e.g. `pulp_timer_unit`. But after reconsidering, I do like `timer_unit` here more. I think its the perfect compromise of specificity and briefness.

File: examples/headsail-bsp/src/mmap.rs
Comment (hegza) [inline examples/headsail-bsp/src/mmap.rs:2]: Thanks for adding the underscore here. I think it's generally a good idea to include those in the addresses.

File: examples/headsail-bsp/src/timer_unit.rs
Comment (hegza) [inline examples/headsail-bsp/src/timer_unit.rs:9]: Excellent header comment. This makes it very clear that this is VP only.

I think the date is a bit superfluous, as git already takes care of that for us, and there will be changes to this driver later which will muddy the waters with that anyway.

File: examples/headsail-bsp/src/timer_unit.rs
Comment (hegza) [inline examples/headsail-bsp/src/timer_unit.rs]: This `extern crate` is 2015 syntax, and is no longer necessary in the 2023 edition. Dependency crates are imported implicitly on first use. You may leave this out and the program will be  equivalent. Less lines is better.

File: examples/hpc/hello-hpc/examples/timer0.rs
Comment (hegza) [inline examples/hpc/hello-hpc/examples/timer0.rs:1]: This program uses the timer0_get_count which is only available in debug builds, but does not signal this in any way. I would prefer all programs can be compiled in `release` and `debug`, allowing for `cargo build --examples` to work in all  invocations without edge cases. Therefore, I recommend that you conditionally compile out the lines that use the debug-only features.

You can test everything is ok by running `cargo build --example timer0` and `cargo build --release --example timer0` and seeing that both complete without errors.

Comment (hegza) [reply examples/hpc/hello-hpc/examples/timer0.rs:1]: Alternatively, making timer0_get_count available in release builds looks like a valid option here

File: examples/headsail-bsp/src/timer_unit.rs
Comment (hegza) [inline examples/headsail-bsp/src/timer_unit.rs]: The `TIMER0_ENABLE_BIT` constant is used as an indexer in all contexts, therefore its type should probably be expressed as `usize`. This will also allow you to remove the redundant type casts below.

File: examples/headsail-bsp/src/timer_unit.rs
Comment (hegza) [inline examples/headsail-bsp/src/timer_unit.rs]: `reg as u32` this looks like a redundant type conversion. We have already read a u32 from the register and we're putting a u32 back in. A type conversion should not be necessary.

**Review by hegza — CHANGES_REQUESTED** [review]
Looks quite alright with minor changes. We might want to upgrade it to a relocatable HAL driver later, but that's a tomorrow's problem we don't need to be concerned with now.

Congratulations on your first Rust driver :smiley:

File: examples/headsail-bsp/src/lib.rs
Comment (hegza) [inline examples/headsail-bsp/src/lib.rs:6]: One more thing! We might want to conditionally compile this line with `[cfg(feature = "vp")]` to make sure it's not included when targeting the ASIC. Then add the line `vp = []` under `[features]` in Cargo.toml, to declare the feature.

Comment (hegza) [reply examples/headsail-bsp/src/lib.rs:6]: Ah, it's in the other PR. I would've put it in this one already but that's all fine. Consider this comment chain void.

File: examples/headsail-bsp/Cargo.toml
Comment (andstepan) [reply examples/headsail-bsp/Cargo.toml:58]: Done, moved timer0.rs to headsail-bsp/examples.

File: examples/headsail-bsp/src/timer_unit.rs
Comment (andstepan) [reply examples/headsail-bsp/src/timer_unit.rs]: Done, removed line 12.

File: examples/hpc/hello-hpc/examples/timer0.rs
Comment (andstepan) [reply examples/hpc/hello-hpc/examples/timer0.rs:1]: Yes, I agree with your second comment, I think timer0_get_count() should probably be for release too as it can be used to profile execution times etc. However, the timer0_get_ctrl_reg() should probably remain debug only as I see no reason why would someone want to read the timer's control register during runtime. The user of the driver will probably just configure the timer and use it, never reading back what is in the control register.
I removed the debug_assertions from the timer0_get_count() function.

File: examples/headsail-bsp/src/timer_unit.rs
Comment (andstepan) [reply examples/headsail-bsp/src/timer_unit.rs]: My initial idea was that this should be u3 instead of usize since we are referring to registers and not pointers and those registers are always 32 bits wide. I have corrected the code, but it is still not very clear to me why usize should be used here instead of u32.

File: examples/headsail-bsp/src/timer_unit.rs
Comment (andstepan) [reply examples/headsail-bsp/src/timer_unit.rs]: Done, removed type conversion.

File: examples/headsail-bsp/src/timer_unit.rs
Comment (hegza) [reply examples/headsail-bsp/src/timer_unit.rs]: One superfluous cast still remains on line 39 but we can fix that later.

Comment (hegza) [reply examples/headsail-bsp/src/timer_unit.rs]: I think what is meant here with this constant is "which bit is our target bit, of the 32 options in the u32", with a value range of `0..32` (so technically u8 suffices if we wanted to optimize for space; though we do not).

We are using the value to do a bit-shift (even if that bit shift is encapsulated by the `bit_field` library) and the bit-shift operation takes a usize as a parameter. This is reflected in the API of the `bit_field` library which also takes a `usize` as the parameter in the `set_bit` function.

It's indeed typical that when a number is used as an indexer in a table such as an array of bits, it uses the native platform pointer width a.k.a. `usize`.

**Review by hegza — APPROVED** [review]
Looks good to me now. I'll let you press the squash & merge button in case there's something you'd like to change prior.

**Pull request merged**
